src/client/Client.cc

```diff
@@ -6162,30 +6162,37 @@ int Client::link(const char *relexisting, const char *relpath, const UserPerm& p
   tout(cct) << relpath << std::endl;
 
   filepath existing(relexisting);
-  filepath path(relpath);
-  string name = path.last_dentry();
-  path.pop_dentry();
 
   InodeRef in, dir;
   int r = path_walk(existing, &in, perm, true);
   if (r < 0)
     goto out;
-  r = path_walk(path, &dir, perm, true);
-  if (r < 0)
+  if(std::string(relpath) == "/") {
+    r = -EEXIST;
     goto out;
-  if (cct->_conf->client_permissions) {
-    if (S_ISDIR(in->mode)) {
-      r = -EPERM;
-      goto out;
-    }
-    r = may_hardlink(in.get(), perm);
-    if (r < 0)
-      goto out;
-    r = may_create(dir.get(), perm);
+  }
+  {
+    filepath path(relpath);
```

Comment:
If I undo this, the compiler will complain about goto skipping over definitions of variables.
I initially just wanted to do "return -EEXIST", but since the rest of the function is doing "goto out" I wanted to be consistent.